eth-contracts/contracts/TokenLockerV2.sol

```diff
@@ -0,0 +1,282 @@
+pragma solidity ^0.5.10;
+pragma experimental ABIEncoderV2;
+
+import "./interfaces/IERC20.sol";
+import {CKBCrypto} from "./libraries/CKBCrypto.sol";
+import {TypedMemView} from "./libraries/TypedMemView.sol";
+import {SafeMath} from "./libraries/SafeMath.sol";
+import {CKBTxView} from "./libraries/CKBTxView.sol";
+import {ViewSpv} from "./libraries/ViewSpv.sol";
+import {Address} from "./libraries/Address.sol";
+import {ICKBSpv} from "./interfaces/ICKBSpv.sol";
+import {MultisigUtils} from "./libraries/MultisigUtils.sol";
+
+contract TokenLocker {
+    using SafeMath for uint256;
+    using Address for address;
+    using TypedMemView for bytes;
+    using TypedMemView for bytes29;
+    using CKBTxView for bytes29;
+    using ViewSpv for bytes29;
+
+    uint64 public numConfirmations_;
+    ICKBSpv public ckbSpv_;
+    bytes32 public recipientCellTypescriptCodeHash_;
+    uint8 public recipientCellTypescriptHashType_;
+    bytes32 public bridgeCellLockscriptCodeHash_;
+
+    // txHash -> Used
+    mapping(bytes32 => bool) public usedTx_;
+
+    // refer to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md
+    uint public constant SIGNATURE_SIZE = 65;
+    uint public constant VALIDATORS_SIZE_LIMIT = 20;
+    string public constant name = "Force Bridge TokenLocker";
+    // SET_NEW_CKB_SPV_TYPEHASH = keccak256("SetNewCkbSpv(address newSpvAddress,uint256 nonce)");
+    bytes32 public constant SET_NEW_CKB_SPV_TYPEHASH = 0x27ed7da1c419c8aa5f5c5f74f651895eaf8dab9831123b22248827a707d44bb1;
+    bytes32 public DOMAIN_SEPARATOR;
+    uint public currentSetNewCkbSpvNonce;
+    // if the number of verified signatures has reached `multisigThreshold_`, validators approve the tx
+    uint public multisigThreshold_;
+    address[] validators_;
```

Comment:
The multi-sig admin may be used multiple times, is it possible to abstract it as a contract to inherit or a library?

Reply:
I used modifier before,  it cause “Stack Too Deep” error which often appears in solidity contracts